src/contracts/GPv2Settlement.sol

```diff
@@ -142,4 +150,57 @@ contract GPv2Settlement {
         );
         return IUniswapV2Pair(uint256(pairAddressBytes));
     }
+
+    /// @dev Verifies the specified clearing price is valid. Spefically, this
+    /// method verifies that the ±0.1% fee spread around the clearing price is
+    /// completely within the ±0.3% AMM fee spread around the Uniswap AMM price.
+    /// The AMM price is either defined by the token reserves (AKA spot price)
+    /// in the case the AMM is not traded with, or the actual swap amount (AKA
+    /// effective price) otherwise.
+    function verifyClearingPrice(
+        IUniswapV2Pair pair,
+        int112 d0,
+        int112 d1,
+        uint112 clearingPrice0,
+        uint112 clearingPrice1
+    ) internal view {
+        uint256 uniPrice0;
+        uint256 uniPrice1;
+        if (d0 | d1 == 0) {
+            // NOTE: In case the amounts swapped with the AMM is 0, we want to
+            // ensure that the clearing price with spread is within the AMM
+            // price spread.
+            (uint112 reserve0, uint112 reserve1, ) = pair.getReserves();
+            (uniPrice0, uniPrice1) = (uint256(reserve0), uint256(reserve1));
+        } else {
+            // NOTE: We want to check that exactly only one of `d0` or `d1` is
+            // strictly negative, and the other is strictly positive.
+            // SAFETY: `int112 * int112` cannot overflow a `int256`.
+            require(int256(d0) * int256(d1) < 0, "invalid Uniswap amounts");
```

Comment:
The documentation for overflows is very good. thanks for doing it.

In the past, we concluded that it is beneficial to use safemath multiplications everywhere. While it increases the gas costs a tiny little bit, it gives us peace of mind, in case some future PR will break the safety. It just happens too quickly in some future PR, where one just wanted to "modify this little thingy" something goes wrong.

However, I am not sure whether you can really use safemath here, as safemath operates only in u256 according to my knowledge. If it is not available, then, of course, we can not use it and should just do the "safety explanation".

Reply:
> In the past, we concluded that it is beneficial to use safemath multiplications everywhere. While it increases the gas costs a tiny little bit,

The reason for this is delegate calls are AFAIK 700 gas. So if we use `SafeMath` for each multiplication in this function that is an additional `11 * 700 = 7700` gas just to check the price (`9 * 7700 = 6300` gas if we fix the price to `d0/d1` when we trade with Uniswap).

Additionally, if we start using `SafeMath` for every multiplication when checking limit prices per order for example, then the amount of extra gas needed would grow substantially.

Reply:
>The reason for this is delegate calls are AFAIK 700 gas. So if we use SafeMath for each multiplication in this function that is an additional 11 * 700 = 7700 gas just to check the price (9 * 7700 = 6300 gas if we fix the price to d0/d1 when we trade with Uniswap).

No, if you are inheriting non-public library function, then the code gets directly copied into the original contract and you don't pay the delegatecalls, you are only paying the actual execution. Only for public library function, you have to pay the delegate call costs.

Reply:
>We could cast the uint112 to uint256 and then use SafeMath followed by a SafeCastU112 (which we probably have to implement, but is trivial I think) to achieve this right?

Yeah, but checking whether int112*int112 is negative in uint256 is also non-trivial. In this particular case, probably checking with the math-safe is increasing the complexity a lot, and we can stick to "documentation only". But still, in other places, I guess I would do it. But I am also curious on other opinions.

Reply:
> Only for public library function, you have to pay the delegate call costs.

:exploding_head:. In that case its not an issue.

> int112*int112 is negative in uint256

I hope there is a `SafeMath` with signed operations :crossed_fingers:

Reply:
At this point, wouldn't it be cheaper (and maybe clearer) to write out the full logical expression? `(a<0 && b>0) || (a>0 && b<0)`

Reply:
I don't think it would be cheaper since each `&&` potentially jumps, but it would be clearer than `SafeMath`.